File: Python/2017/Day23/main.py
from collections import defaultdict
from copy import deepcopy

with open(0) as f:
    instructions = [x.strip() for x in f.readlines()]

instructions = [i.split() for i in instructions]

def run(a):
    def get(x):
        try:
            x = int(x)
            return x
        except:
            return R[x]

    R = defaultdict(int)
    H = []
    R['a'] = a
    mul = 0
    i = 0
    J = defaultdict(lambda: int(1e9))
    count = 0

    while i < len(instructions):
        cmd, x, y = instructions[i]

        if cmd == "set":
            R[x] = get(y)
        elif cmd == "sub":
            R[x] -= get(y)
        elif cmd == "mul":
            R[x] *= get(y)
            mul += 1
        elif cmd == "jnz":
            if get(x) != 0:
                i += get(y)
                continue
        else:
            assert False, (cmd, x, y)

        if x == 'h':
            H.append(R['h'])

        count += 1
        i += 1

    return mul if a == 0 else R['h']

print("Part 1:", run(0))
print("Part 2:", run(1))

File: Rust/2017/day23/src/main.rs
#![allow(unused)]

use std::io::{self, Read};
use std::vec;
use std::io::{Write, BufReader, BufRead, ErrorKind};
use std::fs::File;
use std::cmp::Ordering;
use std::collections::{HashMap, HashSet};

fn read_file(file_name: &str) -> String {
    let mut file = File::open(file_name).unwrap();
    let mut contents = String::new();
    file.read_to_string(&mut contents).unwrap();
    contents
}

fn run(a: i64, instructions: &Vec<Vec<&str>>) -> i64 {
    let mut registers: HashMap<char, i64> = HashMap::new();
    let mut i = 0;
    let mut mult = 0;

    fn get(x: &str, registers: &HashMap<char, i64>) -> i64 {
        return match x.parse() {
            Ok(x) => x,
            Err(_) => **registers.get(&x.chars().nth(0).unwrap()).get_or_insert(&0i64),
        }
    }

    while i < instructions.len() {
        let parts = instructions[i];

        match parts[0] {
            "set" => registers[&parts[1].chars().nth(0).unwrap()] = get(parts[2], &registers),
            "sub" => registers[&parts[1].chars().nth(0).unwrap()] -= get(parts[2], &registers),
            "mul" => registers[&parts[1].chars().nth(0).unwrap()] *= get(parts[2], &registers),
            "jnz" => {
                let val = registers.get(&parts[1].chars().nth(0).unwrap()).get_or_insert(&0i64);
                if **val != 0 {
                    i += get(parts[2])
                }
            },
            _ => panic!("OH NO!"),
        }
    }

    mult
}

fn main() {
    let input = read_file("input.txt");

    let instructions = input.lines().map(|x| x.split(' ').collect::<Vec<_>>()).collect::<Vec<_>>();


}

